include/seahorn/Expr/TypeChecker.hh

```diff
@@ -1,17 +1,66 @@
 #pragma once
-
-#include "seahorn/Expr/ExprApi.hh"
+#include "seahorn/Expr/ExprCore.hh"
+#include "seahorn/Expr/ExprOpSort.hh"
 
 namespace expr {
-class TypeChecker {
+
+class TypeChecker;
+
+class TypeCheckerHelper {
+  friend class TypeChecker;
+
   class Impl;
-  Impl* m_impl;
+  Impl *m_impl;
+
+  TypeCheckerHelper();
+  ~TypeCheckerHelper();
+
+public:
+  Expr typeOf(Expr e);
+  Expr getErrorExp();
+
+  /*
+  Used to keep track of which expressions are bound variables
+  */
+  void mapBoundVar(Expr bVar);
+
+  /*
+  Returns set of all the bound variables that the expressions has. This includes
+  any bound variables in any of its sub expressions
+  */
+  ExprSet getBoundVars(Expr exp);
+
+  /*
+  Used to keep track of which expressions are binders
+  */
+  void mapBinder(Expr binder);
+};
+
+class TypeChecker {
+  TypeCheckerHelper m_helper;
 
 public:
   TypeChecker();
-  ~TypeChecker();
+
+  /*
+  See typeOf(Expr e);
+  */
   Expr sortOf(Expr e) { return this->typeOf(e); }
-  Expr typeOf(Expr e); 
+
+  /*
+  Returns the type of the passed expression. If an error is found (ie. the
```

Comment:
Use doxygen style in these user facing comments. We don't really run doxygen, but it helps in structuing the comments and referring to argument names and returns. Light doxygen is ok